server/src/main/java/io/crate/expression/symbol/ScopedSymbol.java

```diff
@@ -54,11 +57,19 @@ public final class ScopedSymbol implements Symbol {
     private final RelationName relation;
     private final ColumnIdent column;
     private final DataType<?> dataType;
+    @Nullable
+    private final Symbol symbol;
 
-    public ScopedSymbol(RelationName relation, ColumnIdent column, DataType<?> dataType) {
+    public ScopedSymbol(RelationName relation, ColumnIdent column, DataType<?> dataType, @Nullable Symbol symbol) {
         this.relation = relation;
         this.column = column;
         this.dataType = dataType;
+        // the original 'symbol' is required to be able to differentiate ScopedSymbols with identical ColumnIdents:
+        // ex) select * from (select t1.a, t2.a from t as t1, t as t2) t3;
+        // The aliasedAnalyzedRelation for 't3' used to have two of 'ScopedSymbol(t3, a, int)',
+        // but now it has 'ScopedSymbol(t3, a, int, ScopedSymbol(t1, a, int))' and 'ScopedSymbol(t3, a, int, ScopedSymbol(t2, a, int))'.
+        // NOTE: symbol.valueType() != this.dataType, see now UnionSelect instantiates ScopedSymbols.
```

Comment:
This should be better placed at the ctors java doc, documenting the `symbol` parameter.
It's quite verbose, I think mentioning that 2 ScopedSymbols relating to different source symbols cannot be differentiated later on and thus may wrongly get pruned.

Reply:
Thanks for reviewing, since it is suggested to avoid modifying `ScopedSymbol` I am resolving all reviews regarding the old commit for now.